tests/system_tests_tcp_adaptor.py

```diff
@@ -1045,11 +1090,16 @@ def run_ncat(self,
                  timeout=10,
                  data=b'abcd',
                  use_ssl=False,
-                 use_client_cert=False):
+                 use_client_cert=False,
+                 use_bad_ca_cert=False,
+                 error_ok=False):
```

Comment:
Is this necessary?  Why not let run_ncat raise an exception on failure and have the tests use self.assertRaises() in those cases where run_ncat is expected to fail?

Reply:
I don't like the `raise Exception()`. It would be nicer to create a subclass and throw that. Also, you can create fields on the subclass, so the results are easier to pull out

```python
TestProcessException(Exception):
    def __init__(self, return_code, std_out, std_err):
        self.return_code=return_code
        ...
```

Or maybe create parameter `throw_on_falure=True`, and when that's false, return the result of execution as return value and avoid catching expected exceptions? (I am not fan of `assertRaises()`, although there is nothing wrong with using it, of course.)

Reply:
There already is https://docs.python.org/3.6/library/subprocess.html#subprocess.CalledProcessError, maybe just catch that?